packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,423 @@
+   * @param {StreamCell['values']} values
+   */
+  function* allUpdatesFromCell(blockHeight, values) {
+    for (const value of values)
+      yield /** @type {Update<T>} */ ({
+        blockHeight,
+        value,
+      });
+  }
+
+  /**
+   * @param {Height} [minimum]
+   * @param {Height} [maximum]
+   */
+  async function* iterate(minimum, maximum) {
+    /** @type {Array<Update<T>>} */
+    let values = [];
+
+    for await (const value of reverseIterate(maximum, minimum))
+      values = [value, ...values];
+
+    for (const value of values) yield value;
+  }
+
+  /**
+   * @param {Height} height
+   */
+  const latest = async height => {
+    if (height && height < MINIMUM_HEIGHT)
+      throw makeVstorageError({
+        errorMessage: `${INVALID_HEIGHT_ERROR_MESSAGE} ${height} is less than minimum height ${MINIMUM_HEIGHT}`,
+        height,
+        kind: PATHS.DATA,
+        path,
+      });
+
+    const response = await queryClient.queryAbci(path, {
+      height,
+      kind: PATHS.DATA,
+    });
+
+    const parsedValue = /** @type {StreamCell} */ (parseValue(response.value));
+    assertValueStructure(parsedValue);
+
+    return /** @type {Update<T>} */ ({
+      blockHeight: BigInt(parsedValue.blockHeight),
+      value: parsedValue.values.reverse().find(Boolean),
+    });
+  };
+
+  /**
+   * @param {Height} [maximum]
+   * @param {Height} [minimum]
+   */
+  async function* reverseIterate(maximum, minimum) {
+    let blockHeight = maximum;
+
+    await null;
+
+    if (!minimum || minimum < MINIMUM_HEIGHT) minimum = MINIMUM_HEIGHT;
+
+    if (maximum && maximum < minimum)
+      throw makeVstorageError({
+        errorMessage: `${INVALID_HEIGHT_ERROR_MESSAGE} maximum height ${maximum} is less than minimum height ${minimum}`,
+        height: blockHeight,
+        kind: PATHS.DATA,
+        path,
+      });
+
+    while (true) {
+      /** @type {string | undefined} */
+      let value;
+
+      try {
+        ({ value } = await queryClient.queryAbci(path, {
+          height: blockHeight,
+          kind: PATHS.DATA,
+        }));
+      } catch (err) {
+        console.error(err);
+        break;
+      }
+
+      if (!value) break;
+
+      const parsedValue = /** @type {StreamCell} */ (parseValue(value));
+      assertValueStructure(parsedValue);
+
+      const currentBlockHeight = BigInt(parsedValue.blockHeight);
+
+      if (currentBlockHeight >= minimum)
+        yield* allUpdatesFromCell(
+          currentBlockHeight,
+          parsedValue.values.reverse(),
+        );
+
+      if (currentBlockHeight <= minimum) break;
+
+      blockHeight = currentBlockHeight - 1n;
+    }
+  }
+
+  return {
+    iterate,
+    latest,
+    reverseIterate,
+  };
+};
+
+/**
+ * @param {{fetch: typeof window.fetch}} powers
+ * @param {MinimalNetworkConfig} config
+ */
+export const makeVStorageClient = ({ fetch }, config) => {
+  if (!config.rpcAddrs.length)
+    return Fail`${INVALID_RPC_ADDRESS_ERROR_MESSAGE} ${quote(config)}`;
```

Comment:
The standard pattern is ```$condition || Fail`…` ```:
```suggestion
  config.rpcAddrs.length ||
    Fail`${INVALID_RPC_ADDRESS_ERROR_MESSAGE} ${quote(config)}`;
```
but I think that in general, we're preferring less granular configuration as seen at e.g. https://github.com/Agoric/agoric-sdk/blob/de4f3d31a32806c1805c35018ca98e0e36b4c842/packages/agoric-cli/src/commands/wallet.js#L33 with use of an `AGORIC_NET` environment variable that is mapped by `fetchEnvNetworkConfig` to a https://${AGORIC_NET}.agoric.net/network-config URL (cc @mhofman).